sdk/core/azure-core/test/ut/transport_adapter_curl.cpp

```diff
@@ -36,89 +37,131 @@ namespace Azure { namespace Core { namespace Test {
     }
   } // namespace
 
-/***********************  Unique Tests for Libcurl   ********************************/
-//  Disabling test with INCLUDE_DISABLED_TESTS. The test name cannot be changed because it depends
-//  on a friend class definition. Hence, it can't use the gtest DISABLE_.
-#if defined(INCLUDE_DISABLED_TESTS)
-    TEST_P(TransportAdapter, connectionPoolTest)
+  /***********************  Unique Tests for Libcurl   ********************************/
+  TEST_P(TransportAdapter, connectionPoolTest)
+  {
+    Azure::Core::Http::CurlConnectionPool::ClearIndex();
+
+    // Make sure there are nothing in the pool
+    EXPECT_EQ(Azure::Core::Http::CurlConnectionPool::ConnectionPoolIndex.size(), 0);
+
+    // Use the same request for all connections.
+    Azure::Core::Http::Request req(
+        Azure::Core::Http::HttpMethod::Get, Azure::Core::Http::Url("http://httpbin.org/get"));
+
     {
-      Azure::Core::Http::Url host("http://httpbin.org/get");
-      Azure::Core::Http::CurlConnectionPool::ClearIndex();
-
-      auto threadRoutine = [&]() {
-        using namespace std::chrono_literals;
-        auto request = Azure::Core::Http::Request(Azure::Core::Http::HttpMethod::Get, host);
-        auto response = m_pipeline->Send(Azure::Core::GetApplicationContext(), request);
-        checkResponseCode(response->GetStatusCode());
-        auto expectedResponseBodySize = std::stoull(response->GetHeaders().at("content-length"));
-        CheckBodyFromBuffer(*response, expectedResponseBodySize);
-        std::this_thread::sleep_for(1s);
-      };
-
-      std::thread t1(threadRoutine);
-      std::thread t2(threadRoutine);
-      t1.join();
-      t2.join();
-
-      // 2 connections must be available at this point
-      EXPECT_EQ(Http::CurlConnectionPool::ConnectionsOnPool("httpbin.org"), 2);
-
-      std::thread t3(threadRoutine);
-      std::thread t4(threadRoutine);
-      std::thread t5(threadRoutine);
-      t3.join();
-      t4.join();
-      t5.join();
-
-      // Two connections re-used plus one connection created
-      EXPECT_EQ(Http::CurlConnectionPool::ConnectionsOnPool("httpbin.org"), 3);
+      // Creating a new connection with default options
+      Azure::Core::Http::CurlTransportOptions options;
+      auto session = std::make_unique<Azure::Core::Http::CurlSession>(
+          req,
+          Azure::Core::Http::CurlConnectionPool::GetCurlConnection(req, options),
+          options.HttpKeepAlive);
+      // Simulate connection was used already
+      session->m_lastStatusCode = Azure::Core::Http::HttpStatusCode::Ok;
+      session->m_sessionState = Azure::Core::Http::CurlSession::SessionState::STREAMING;
```

Comment:
When it comes to network, it has been complicated to simulate responses or specific states.
The easiest option is this one where the test can exactly set the next expected state and test some outcome.
The harder option is to add mocked objects where, instead of the test setting this state, we have to model what the mocked object will do.

Note that in both cases, the way we set the specific state is 100% tight coupled to the implementation. That's just how it is XD
Unit tests are meant to test your implementation. If you change your impl, you need to change your tests.
Functional tests are meant to test your public surface. Only when public surface change, the test needs to change